### PR TITLE
Use snapshot data from cache instead of resnapshotting in NextGenBuildCacheController

### PR DESCRIPTION
**Pull request by asodja**

### CODE REVIEW

File: subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/NextGenBuildCacheController.java [hunk 1]
```diff
@@ -216,6 +233,64 @@ public ImmutableSortedMap<String, FileSystemSnapshot> getResultingSnapshots() {
         });
     }
 
+    private FileSystemLocationSnapshot getSnapshot(File root, List<ManifestEntry> entries) {
```
Comment (lptr) [inline subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/NextGenBuildCacheController.java:236]: 🧐 Maybe not call these methods `get.*Snapshot()` but `create.*Snapshot()`?

File: subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/NextGenBuildCacheController.java [hunk 2]
```diff
@@ -216,6 +233,64 @@ public ImmutableSortedMap<String, FileSystemSnapshot> getResultingSnapshots() {
...
+        ManifestEntry rootEntry = entries.get(0);
+        switch (rootEntry.getType()) {
+            case Directory:
+                return getDirectorySnapshot(root, entries);
+            case RegularFile:
+                File file = new File(root, rootEntry.getRelativePath());
+                return getFileSnapshot(rootEntry, file);
```
Comment (lptr) [inline subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/NextGenBuildCacheController.java:243]: ❓ When the whole tree is a single file, then `root` should already point to the file, no?

File: subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/NextGenBuildCacheController.java [hunk 3]
```diff
@@ -31,20 +32,26 @@
...
 import org.apache.commons.io.output.TeeOutputStream;
+import org.gradle.api.internal.cache.StringInterner;
 import org.gradle.caching.BuildCacheEntryWriter;
 import org.gradle.caching.BuildCacheKey;
 import org.gradle.caching.internal.CacheableEntity;
 import org.gradle.caching.internal.DefaultBuildCacheKey;
 import org.gradle.caching.internal.controller.CacheManifest.ManifestEntry;
 import org.gradle.caching.internal.controller.service.BuildCacheLoadResult;
 import org.gradle.caching.internal.origin.OriginMetadata;
+import org.gradle.caching.internal.packaging.impl.RelativePathParser;
```
Comment (lptr) [inline subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/NextGenBuildCacheController.java:43]: ❓ Shall we create a copy of this parser, so we can throw out the whole `packaging` package when we switch over completely?

Comment (asodja) [reply subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/NextGenBuildCacheController.java]: We can do that. The original parser was anyway annoying since it required folders to end with '/'. Moved new class as static inner class of NextGenBuildCacheController

Comment (asodja) [reply subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/NextGenBuildCacheController.java]: On second thought, we also have tests for RelativePathParser, so we would need to duplicate that. I will revert that change and use `org.gradle.caching.internal.packaging.impl.RelativePathParser`. I think it would be good we move it to other package once we remove that package.

Comment (lptr) [reply subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/NextGenBuildCacheController.java:44]: OK, your call.

Comment (lptr) [reply subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/NextGenBuildCacheController.java:44]: Please add any follow-up ideas as TODOs in the code so we don't forget.
